esm: implement import.meta.resolve and make nodejs: scheme public

guybedford opened the pull request:
This PR implements `import.meta.resolve` as an async hook into the module resolver.

This fills the gap of enabling `require.resolve` workflows that aren't currently possible in ES modules, such as locating the exact path of a dependency or local resolution following all the correct resolver resolution rules.

This is a feature that will require alignment with browsers, so I'm opening this now to be able to start those discussions. Marking as blocked on consensus discussions.

Usage examples:

* `await import.meta.resolve('./dep.ext')` -> `file:///path/to/dep.ext`, relative to current module.
* `await import.meta.resolve('./dep.ext', 'file:///different/path/parent')` -> `file:///different/path/dep.ext`, relative to provided parent.
* `await import.meta.resolve('dependency')` -> `file:///path/to/node_modules/dependency/main.js`, as resolved from current module.
* `await import.meta.resolve('dependency', 'file:///different/path/parent')` -> `file:///different/path/node_modules/dependency/main.js`, as resolved from provided module.
* `await import.meta.resolve('./does-not-exist')` -> Throws not found as the default Node.js resolver checks for existence.
* `await import.meta.resolve('fs')` -> `nodejs:fs`

If the Node.js resolver has been hooked by a loader, the corresponding resolve function will be called in `import.meta.resolve`.

This API also makes the `nodejs:fs` internal builtin module scheme public, including the associated changes for this feature.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

guybedford commented:
//cc @nodejs/modules-active-members

devsnek commented on lib/internal/modules/esm/translators.js:
```diff
@@ -63,16 +64,27 @@ function errPath(url) {
+  return esmLoader.import(specifier, url);
+}
+
+async function importMetaResolve(specifier, parentUrl = this.url) {
```
We should definitely document where `this.url` is coming from (`import.meta`). I think I'd also prefer not depending on the receiver at all, since `const { xyz } = import.meta` is probably not going to be a wholly uncommon pattern.

guybedford replied:
Sure, I've fixed this up to use a proper closure. The idea of a function closure per module still abhors me, but I will get over it.

guybedford commented:
Discussion issue: https://github.com/nodejs/modules/issues/458

guybedford commented:
@devsnek as mentioned in the issue comment, this PR is to start exactly those discussions, and is blocked on them.

devsnek commented:
@guybedford oh i see, I started writing my review before you posted that comment :) Glad to see we're on the same page :+1:

jkrems commented:
> `await import.meta.resolve('./does-not-exist')` -> Throws not found as the default Node.js resolver checks for existence.

Can we prevent that? It feels constraining when `resolve` has to know already if the URL can be loaded. E.g. we could return the resolution as if path extensions wouldn't be tried..? Basically I want to prevent that it becomes observable in node that `.resolve('./foo')` can actually check for foo's existence because the same code running elsewhere wouldn't realistically provide that info.

ljharb commented:
`require.resolve` already checks for existence and throws - while I'm not a fan of an exception there, the existence check is actually quite important for all the use cases of a resolver I'm aware of.

jkrems commented:
If you're fine actually touching the file, it feels like you could just `import` it if you want to check that it's in fact reachable? E.g. throwing implies that if it doesn't throw, the file can be loaded. Which isn't necessarily true:

```
$ cat x.js
cat: x.js: Permission denied
$ node -e "require.resolve('./x.js')"
$ node -e "require('./x.js')"
internal/fs/utils.js:220
    throw err;
    ^

Error: EACCES: permission denied, open '/private/tmp/x.js'
```

It feels like it encourages people to assume that just because they could resolve it, there's some sort of guarantee that the returned string can be loaded.

guybedford commented:
If we want to move the existence check from the resolver to the instantiation phase, we can do that, but that should be considered a separate concern to this PR. Also someone would need to justify and create that PR before stability.

jkrems commented:
Right now we don't leak when exactly the exception happens to `import`, so it's just an implementation detail. I would consider leaking it a compat risk with the web since even if the web adds `import.meta.resolve`, I assume it won't be supporting existence checks. If we get signals from the web that they will include existence, then that changes this whole thing. :)

ljharb commented:
@jkrems it is definitely not an option to `import()` a file, and risk evaluating code, to check for existence - that's a big part of the point of a resolver, to know it will work before trying it.

jkrems commented:
Right, but that’s not actually possible to derive from the resolution result. As shown above, file permissions are enough for a resolved path not to be valid for import. And this doesn’t touch things like race conditions or loaders generating files on-the-fly (or hiding them even though they exist on disk).

What is a use case where you need to know that you could require something but where you can’t try to actually require it?

ljharb commented:
That there are more reasons why an import might not succeed (including file permissions, transitive imports failing, exceptions) doesn't change that "the specifier doesn't resolve to something" is an important bit to know prior to attempting to evaluate it. One example in particular, this bit helps you catch a programmer error on the importing side using eslint-plugin-import, which is used in eslint-config-airbnb.

jkrems commented:
> One example in particular, this bit helps you catch a programmer error on the importing side using eslint-plugin-import, which is used in eslint-config-airbnb.

I know (and actively use) that plugin. But if that plugin wanted to start using `import.meta.resolve`, couldn't it just `stat` the resulting file to see if it's there? It could then even fail if it wasn't readable etc..

My primary problem with leaking it is quite simple: It means that `import.meta.resolve` is very unlikely to be portable for minimal gain. I'd be surprised if browser vendors would be willing to roll a resource existence / reachability check into a function to resolve URLs. This smells like `URL.equals` doing DNS queries in Java.*

* Yes, node's default resolver accesses the file system. But one can implement the ES resolver using a "small" in-memory data structure that doesn't require knowledge about every single file on disk (an import map). If file existence is required to implement resolution, import maps are no longer sufficient to model import URL resolution in node. To me that's a big property to give up on for a pretty fuzzy check of "could this URL actually be fetched".

jkrems commented:
For illustration of why I wouldn't want `resolve` to throw in general: Adding a step that tries to access the resource is easy. Undoing that exception is *hard*: https://github.com/jkrems/node/commit/6cebcb20c27e6bf7bd52f5ac9afd6923f6e6e851#diff-6dda3ef5f860111452d324058e12a738R162-R183

ljharb commented:
How would your alternative handle different kinds of exceptional cases, like in that snippet? How would the API return a success or multiple kinds of error values on the same channel, and why would that be an improvement?

jkrems commented:
I assume the only actual error case you'd likely be interested in would be that the file doesn't exist. And that can be checked by stating it. Without running the entire loader hooks chain (including source hooks and/or transforms), we won't be able to return any of this info longer term if loader hooks are registered. So there would be a `resolve` function that loads files from disks and compiles them. Which just seems wrong.

I'm happy to support a `async isLoadableAsCode(url): boolean` API in addition to this that would do all these checks. But I want to be able to resolve to a directory. Or to a text file I'll never import. So even ignoring the weird side effects to the implementation of `.resolve`, it feels like an awkward API to have opinions on what the returned URL is used for.

ljharb commented:
If i have to manually stat it, then “there’s a filesystem” isn’t abstracted from users of the resolver. One of the big benefits is to abstract that away. I shouldn’t need any fs methods at all to resolve a specifier, just the abstraction.

ljharb commented on doc/api/esm.md:
````diff
@@ -808,6 +808,31 @@ const __filename = fileURLToPath(import.meta.url);
+})();
+```
+
+`import.meta.resolve` also accepts a second argument which is the parent module
````
require.resolve doesn’t do this; perhaps we could make it do this, so the two APIs are parallel?

ExE-Boss replied:
It’s been supported since <https://github.com/nodejs/node/pull/16397> as:
```js
require.resolve('./dep', { paths: [ __filename ] });
```

jkrems commented:
> If i have to manually stat it, then “there’s a filesystem” isn’t abstracted from users of the resolver. One of the big benefits is to abstract that away. I shouldn’t need any fs methods at all to resolve a specifier, just the abstraction.

I think I acknowledged that when I suggested that there could be a second function (`isLoadableAsCode`) that could do all the resource retrieval & content type testing required for more certainty. I just don't think that restricting `import.meta.resolve` to just this one case (only allow resolving to files that could likely be loaded via import as code) is worth it.

Especially because `import.meta.resolve` will unlikely be used by tools like `eslint-plugin-import` - because I assume for that plugin you wouldn't want to create an actual ES module with a faked URL inside of a node instance that has the proper content types supported you would need for the configuration of the plugin.

devsnek commented:
i think this argument may be a bit too isolated on the specific semantics of resolve. taking a step back, we have a pipeline somewhat like this:

`import() => (specifier, referrer) => resolve() => url => getSource() => ...`

The URL given to `getSource()` needs to be final. that is, `getSource()` understands how to load a found resource, but it does not understand how to find that resource.

Any behaviour needed to go from `(specifier, referrer)` to `url` should therefore be performed during `resolve()`. This may be as simple as `new URL(specifier, referrer).toString()` or it may be more complex, such as what our default resolver does.

The constraints of this model dictate that it *must* be assumed that `resolve()` can fail. For example, in our default resolver, missing permissions may preclude the read operation on a package.json, which means the algorithm doesn't know where to look for the entry file. The only reasonable behaviour is to bail.

jkrems commented:
> For example, in our default resolver, missing permissions may preclude the read operation on a package.json, which means the algorithm doesn't know where to look for the entry file. The only reasonable behaviour is to bail.

I think that's fair. There's definitely exceptional circumstances where resolution cannot determine what the final `url` would be. I think throwing errors in that case would be the right thing to do.

guybedford commented:
I've added a new commit here with the following features:

1. `resolve` is now separated from `getFormat`, so `import.meta.resolve('./file.unknown')` resolving an unknown extension will not call `getFormat` and throw - it will just do the resolve operation and return the resolved file.
2. I've added a special case support for a trailing slash form - `import.meta.resolve('./dir/')` which will always return the directory path retaining the trailing slash. This trailing slash then effectively opts out of the file existence check (although the directory should exist).

With the above a lot of the nice browser use cases for import.meta.resolve are kept despite the existence check. For example resolving the URLs of asset files and folders within packages.

Happy to still discuss the moving the existence check out of the resolver, but I do feel that Node.js resolver consistency is very much based on existence, especially given that we always check existence with the `realpath` call anyway, so that separating existence again would even be a minor perf concern.

jkrems commented:
> Happy to still discuss the moving the existence check out of the resolver, but I do feel that Node.js resolver consistency is very much based on existence, especially given that we always check existence with the realpath call anyway, so that separating existence again would even be a minor perf concern.

To me it's less about the performance and more about consistency. This will mean that we do existence checks for file URLs but not for https URLs. And it will mean that a static resolver *may* change the observable behavior of resolve. If we're fine with that, I can live with shipping this. I assume in the 99% case it will be used to resolve against an existing target or a directory. 🤞

BridgeAR commented:
This needs a rebase

Jamesernator commented:
> To me it's less about the performance and more about consistency. This will mean that we do existence checks for file URLs but not for https URLs.

I for one would definitely prefer not to see existence checks, as I often use relative urls from `import.meta` for things like caches or logs that may not exist on first run.

e.g. Ideally this would still work:

```js
const logger = new Logger(await import.meta.resolve(`./out.log`));
const processedImageCache = new ValueCache(await import.meta.resolve(`./caches/processed-images/`));
```

devsnek commented:
resolve doesn't validate the importability of a specifier, it just finds the absolute URL. that may invoke certain checks, but checking that the resulting resource exists is not one of them.

ljharb commented:
That’s what path.resolve does. require.resolve checks existence. import.meta.resolve is akin to require.resolve, not path.resolve.

guybedford commented:
Per the meeting discussion, I've updated this PR to be flagged behind `--experimental-import-meta-resolve`.

guybedford commented:
Since this PR also effectively exposes the ability for `import.meta.resolve('fs')` to return `node:fs`, which up until now has been a private internal scheme, I've gone ahead and also renamed this scheme to `nodejs:` and made the other changes necessary to treat it as a public scheme (such as ensuring resolve always returns this URL, and allowing `import.meta.resolve('nodejs:fs')` to work too).

ljharb commented:
Why not "node"? "nodejs" is only what ubuntu's default distro mistakenly calls node because node is used by a ham radio app.

guybedford commented:
Previous discussions on this topic have skewed towards a `nodejs:` consensus, but I'm open to either personally.

jkrems commented:
Approved for overall API. I expect that some details may have to change as the exact details of the loader hooks get figured out. But that can happen while it's experimental.

ljharb commented:
Hopefully this will start behind an explicit flag instead of being unflagged and experimental?

jkrems commented:
Yes, it's currently behind `--experimental-import-meta-resolve`.

addaleax commented on doc/api/cli.md:
```diff
@@ -156,6 +156,13 @@ Enable experimental Source Map V3 support for stack traces.
 
+### `--experimental-import-meta-resolve`
+<!-- YAML
+added: v13.x.x
```
```suggestion
added: REPLACEME
```

MylesBorins commented:
What else is needed for us to land this?

jkrems commented:
> What else is needed for us to land this?

I think this is ready to land unless Guy protests.

guybedford commented:
Landed in https://github.com/nodejs/node/commit/0f96dc266fd0cd8c1baa82ce7eb951c11b29a331.

ExE-Boss commented:
The signatures of `import.meta.resolve(…)` and `require.resolve(…)` differ in terms of which arguments they accept: <https://github.com/nodejs/node/issues/31861>.
